### ISSUE TITLE
[RFC] consider using TOML over YAML

### ISSUE THREAD

**Issue by derekdreery** [amethyst/amethyst#171]
Using TOML for config files would match rust's own config, and it is significantly easier to write (IMHO) than YAML/JSON.

https://npf.io/2014/08/intro-to-toml/

[TOML](https://github.com/alexcrichton/toml-rs) also has backends for [serde](https://github.com/serde-rs/serde) and rustc-serialize if you want to deserialize into custom types.

**Comment by derekdreery** [comment]
I'm new to this project so I may have misunderstood the use of yaml in the project_template directory.

**Comment by ebkalderon** [comment]
Hi there @derekdreery, and welcome! I originally chose YAML because it is much more ergonomic when working with deeply nested hierarchies, unlike TOML. While I agree with you that TOML is super nice for writing configuration files, it's not so nice for building game entities and describing scene hierarchies. YAML is less adept at this, but is still reasonably easy to write, edit, and understand.

I briefly considered having both formats present, with YAML for entities/scenes and TOML for engine configuration, but decided against it since forcing the user to learn yet another language to use Amethyst (i.e. Rust, YAML, TOML, and a gameplay scripting language) would just frustrate potential new users for minimal gain.

Hope this makes sense!

**Comment by derekdreery** [comment]
Hey, yeah makes sense thanks for replying. I definitely agree not to use TOML for graph data!

**Comment by OvermindDL1** [comment]
A question, should Components really ever be deeply nested in a single entity?  Should not one file == one entity?  Thus something like TOML would work quite well.  Or is there a schema for the YAML yet?

_/me is not a fan of YAML, white-space sensitivity can be a huge pain..._

**Comment by ebkalderon** [comment]
@OvermindDL1 Yes, I agree in that regard. But descriptions of scene hierarchies, save states, and configuration of more complex stuff like the renderer pipeline and layers are still deeply nested and therefore still need YAML.

**Comment by derekdreery** [comment]
I guess the use case here is *mostly machine read/written, but handy to view for debugging*, where the problem of YAML doesn't stand out so much.

I just never trust a YAML parser because [the spec is so big!](http://yaml.org/spec/1.2/spec.html)

**Comment by derekdreery** [comment]
If it is desirable to have TOML support, it could either be implemented using a *feature* in cargo, or in some sort of pluggable backend at compile-time for these data files. The latter might be good, because then people could also use binary files if they prefered.

**Comment by OvermindDL1** [comment]
Honestly I'd use some simple parser like JSON or S-Expressions or so (hard to be more simple than S-Expressions, if you are not afraid of parenthesis ^.^).

But yeah, the YAML spec is utterly massive for something that should be so simple.  I've been bit by some of its whitespacing stupidity on multiple occasions...

**Comment by Aceeri** [comment]
Supporting multiple formats and going in between them is pretty simple with `serde`, we can just have a simple configuration option for the type of files to use for this (which could just automatically convert them all to that format).

**Comment by torkleyy** [comment]
There is also [ron](https://github.com/kvark/ron). I started a PR for serde support and I'm going finish it by the end of next week.

**Comment by ahmedcharles** [comment]
I'll mention this here since the issue is the result of the difference between YAML and TOML. YAML/JSON both have the concept of an explicit `null`, which the serde implementations map to the `None` case when the type is `Option<T>`. TOML doesn't have an explicit `null`, so the serde implementation doesn't serialize the key. The current deserialization considers missing keys to be worthy of a 'warning' and there's no way to work around the warning. I suspect it's sufficient here to silently go with the default.

**Comment by derekdreery** [comment]
I've kinda gone off using toml since I created this thread, but I find yaml hard to read/write. I guess it's best just to use serde and let the user choose a format, or if that's not possible just use yaml.

**Comment by vitvakatu** [comment]
agreed with @derekdreery

**Comment by torkleyy** [comment]
It seems to me that we should allow the popular options, as mentioned earlier, but still recommend one format that is guaranteed to be compatible with our engine features. 

How does that sound to you?

**Comment by Xaeroxe** [comment]
Gonna be frank: The whole idea is starting to sound a bit over-engineered for how simple the task at hand is.

**Comment by minecrawler** [comment]
Maybe [this blog entry](http://robotlolita.me/2016/01/09/no-i-dont-want-to-configure-your-app.html) is becoming appropriate for the discussion (and similar ones in the future, because they will happen)~

tl;dr: don't give the user too much choice from the beginning. Just make sure it works.

So why don't we just use YAML, say it is the format and people will have to use it? If they get the choice, things will get ugly and more questions will be raised, like "can I have JSON? What about a binary format?" And then Amethyst developers will have to decide on how to design the API in a way that let's the user define which format they chose and so on.

Tbh, as a user of Amethyst, I want to focus on writing a game, not on setting up Amethyst and its millions of choices.

Though it would be nice, indeed, to be able to choose a format (but don't listen to me, guys, I am on Gentoo Linux), the only thing I care about is that I can write/version a config file and Amethyst is able to read the darn config file. Nothing more, nothing less.

**Comment by OvermindDL1** [comment]
I'd still choose about anything except YAML, even ini files...  >.>

**Comment by Xaeroxe** [comment]
I'm a fan of [ron](https://github.com/kvark/ron] which was linked earlier.  Seems like it was purpose made for things like this.  @ebkalderon How would you feel about converting all of our YAML files to this?

**Comment by Aceeri** [comment]
As I said earlier, converting between formats is not hard at all for us to do. With `serde` it is as simple as:
```rust
let mut file = File::open("some_toml.toml")?;
let mut buffer = "".to_owned();
file.read_to_string(&buffer);
let config = toml::from_str::<ExampleConfig>(buffer)?;
config.write("some_yaml_file.yaml")?;
```
Our policy could just be: If serde has a library for the format, sure, otherwise no.

**Comment by Xaeroxe** [comment]
I'd really rather not deal with bugs surrounding interpretations of different config formats.  Can we please Keep It Simple?

**Comment by Aceeri** [comment]
@Xaeroxe We don't have to deal with bugs of different config formats, that is serdes problem, not ours. If someone is try to use some obscure file format that doesn't support half the stuff it probably should for a config format, then they probably shouldn't be trying to use it.

**Comment by Xaeroxe** [comment]
I still don't really understand what gains we get from this.  Why would someone want to use a non-default config file?  It just feels like wasted engineering effort.

**Comment by Aceeri** [comment]
Preferences between config formats as @OvermindDL1 has said they don't really like YAML, plus this has already pretty much been implemented, someone can already use TOML or JSON for their config formats because `Config` implementors implement `Serialize`/`Deserialize` with the macro.

**Comment by Xaeroxe** [comment]
Can we?  Definitely.  Should we?  I don't think so.  

Allowing only one valid file format enables us to load these files with minimum boilerplate, for example I don't need to specify how the file is deserialized, and in some instances I might not even need to specify where the file is.

If an Amethyst tutorial in the future says something like: "Just add this line to your config!" this may or may not work depending on what config format the user is using.

Allowing multiple formats closes more doors than it opens in my opinion.

**Comment by OvermindDL1** [comment]
Never seen RON, I quite like it.

I'm fine with about anything as long as its implementation is fairly bug free and not whitespace significant, YAML fails on both.

Whether JSON, or S-Expressions, or RON, or HOCON (hocon would be awesome, being able to reference definition files from others would be amazing) or whatever all work for me.  Options open is nice, though I can see why only supporting a single one would be useful, I'd still prefer options (I'd take HOCON if I could, hmm, actually I could, there apparently a Rust implementation of it, though maybe not complete...).

**Comment by OvermindDL1** [comment]
For reference, the HOCON spec:  https://github.com/typesafehub/config/blob/master/HOCON.md

**Comment by ebkalderon** [comment]
I agree with @Xaeroxe on this one. We need to endorse a single format and use it everywhere, both in user documentation and in demos. It does not have to be YAML specifically; as long as it is convenient to read or write and works well with nested tables, it should do.

People who write tutorials for Amethyst or open-source their games will vary widely in their choices. If a blog post somewhere shows step-by-step instructions on how to accomplish a task in Amethyst using JSON, but the user's existing config files are written in XML, they will be forced to translate the code snippets into XML with a tool or, worse, by hand before pasting them. It is likely the user will drop either the specific tutorial or Amethyst altogether out of sheer frustration, which is unacceptable.

**Comment by ahmedcharles** [comment]
I didn't intend to incite a flame war. My point was that there's a warning about undefined values that seems inappropriate and whether that can be removed.

**Comment by ebkalderon** [comment]
@ahmedcharles No worries! If anything, I apologize if my comment came off as agitated or reactionary. I just want to take a stance, any stance, on this task since we've been mulling it over for a while now to little effect. :smile:

Your point is quite valid, and to that end, it might just rules out TOML as a contender. Also, since your comment sparked some useful discussion here and on Gitter, there appears to be a growing consensus towards choosing RON to replace YAML, with HOCON coming in as a close second.

**Comment by torkleyy** [comment]
I'd like to use Ron, too (suggested it earlier), and I'm going to finish my PR for it as soon as I'm back home. I'm not sure if @kvark thoght about allowing the same as HOCON does.

**Comment by kvark** [comment]
What exactly do you need from HOCON?

I looked at it briefly, and I saw an attempt to fix existing solutions, which appear inferior to me than building something beautiful from the ground up (thinking Rust versus D as a distant example).

I don't see type-level difference between maps and objects. I see strings that can or can't be quoted. I see thousands of rules and features that don't appear essential to me.

**Comment by torkleyy** [comment]
I was referring to the inclusion of other config files, actually. I'm not
saying it's necessary, just wondering if.

**Comment by Xaeroxe** [comment]
Since we've decided we definitely aren't using TOML I'm closing this, our dicussions on these formats can continue elsewhere.

**Comment by OvermindDL1** [comment]
> What exactly do you need from HOCON?
>
> I looked at it briefly, and I saw an attempt to fix existing solutions, which appear inferior to me than building something beautiful from the ground up (thinking Rust versus D as a distant example).
>
> I don't see type-level difference between maps and objects. I see strings that can or can't be quoted. I see thousands of rules and features that don't appear essential to me.

The main things I like about it are not just the syntax, but primarily the ability to include other HOCON/JSON files at certain points, edit what they contained in their new place, reference values elsewhere to also be here (one-definition is always better than multiple-definition), etc...  It could, for example, let you make a base 'unit', then make an 'attackingunit' that references (includes) unit, maybe makes a change or two, while adding more components, and so forth.  It cuts down on a *substantial* amount of code, plus being able to reference values elsewhere it also means you don't have to worry about some values getting out of sync because you changed them but forgot a place or two.

**Comment by ebkalderon** [comment]
@OvermindDL1 Granted, HOCON's file inclusion feature does seen like a useful tool. But it seems like RON might be a more practical and favored choice long-term. Maybe you could reach out to @kvark and discuss implementing a similar feature in RON over on their repository?

**Comment by Xaeroxe** [comment]
@ebkalderon way ahead of you https://github.com/kvark/ron/issues/11
